Visualize MarginContainer margins when selected

Meorge opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/8961 .

This PR adds lines depicting the margins for a MarginContainer, when it is selected:

<img width="1286" height="786" alt="CleanShot 2025-09-30 at 16 36 32@2x" src="https://github.com/user-attachments/assets/e51e19f7-fadc-42f1-8405-4fc30d5e88a0" />

KoBeWi commented:
This should be done via a dedicated EditorPlugin for MarginContainer instead of hard-coding it in CanvasItemEditor.

Meorge commented:
Ah alrighty, thanks! I'll look into moving the code there 😄

Meorge committed:
Visualize MarginContainer margins when selected

This time as an EditorPlugin!

Improve rendering of inner area of MarginContainer

Decrease thickness and opacity of lines, and render margins as a rectangle rather than lines extending to the edges of the bounding box

Apply suggestions from code review

Co-authored-by: Tomasz Chabora <kobewi4e@gmail.com>

Use `get_margin_size`, change color of viz border, and trigger redraw on MarginContainer's draw signal

AdriaandeJongh commented:
I think this is a fantastic UX improvement, but the PR needs a bit more work I think? can I suggest to make the margin lines not the same color as the bounding box, perhaps a darker or more transparent orange, and have the lines stop at the margin edge instead of at the edge of the control?
<img width="190" height="140" alt="Screenshot 2025-10-01 at 10 13 18" src="https://github.com/user-attachments/assets/0d37d33c-f23e-4766-9703-aa63859acdcd" />

KoBeWi commented on editor/scene/gui/margin_container_editor_plugin.h:
```diff
@@ -0,0 +1,50 @@
+	MarginContainer *margin_container = nullptr;
+
+public:
+	virtual String get_plugin_name() const override { return "MarginContainer"; }
+	bool has_main_screen() const override { return false; }
```
```suggestion
```
Not needed.

KoBeWi commented on editor/scene/gui/margin_container_editor_plugin.cpp:
```diff
@@ -0,0 +1,92 @@
+}
+
+bool MarginContainerEditorPlugin::handles(Object *p_object) const {
+	return p_object->is_class("MarginContainer");
```
```suggestion
	return Object::cast_to<MarginContainer>(p_object) != nullptr;
```

KoBeWi commented:
Modifying margins does not update viewport immediately. You could connect MarginContainer's `draw` signal to `update_viewport()`.

KoBeWi commented on editor/scene/gui/margin_container_editor_plugin.cpp:
```diff
@@ -0,0 +1,92 @@
+
+	Rect2 rect = margin_container->_edit_get_rect();
+
+	int margin_left = margin_container->get_theme_constant("margin_left");
+	int margin_top = margin_container->get_theme_constant("margin_top");
+	int margin_right = margin_container->get_theme_constant("margin_right");
+	int margin_bottom = margin_container->get_theme_constant("margin_bottom");
```
You could use `get_margin_size()` instead.

Meorge replied:
Code-wise, it definitely looks like that - I'll give that a shot!

However, strangely, I don't see any documentation for it. https://docs.godotengine.org/en/latest/classes/class_margincontainer.html
It appears to be a public, non-overriding method, so it seems like it should be documented. Even just searching Google for it doesn't turn up anything: https://www.google.com/search?q=site%3Adocs.godotengine.org+%22get_margin_size%22

Should this be addressed in another PR, or is it intended to be internal to the engine?

KoBeWi commented:
> there is a moment of lag between selecting a MarginContainer and its margins actually being drawn.

If you are using dev build, that's because of inspector being loaded.

KoBeWi reviewed (approved):
Looks fine. Not sure about the yellow color replacing default the selection box (when margins are 0), but it's probably ok.

<img width="407" height="243" alt="Image" src="https://github.com/user-attachments/assets/c65bffe4-43b3-466d-9d9e-29e4d2bee76e" />
